### PR TITLE
Review with quick actions

### PR DESCRIPTION
**Pull request by zampierilucas**
This is the second part of the change to unify the calls made when approving/commenting on an MR.

The new code removes the usage of the `merge_request/approve` API and instead uses the `/approve` and `/unapprove` notes API quick actions.
Fixes: #708

### CODE REVIEW

File: cmd/mr_approve.go
Comment (bmeneguele) [inline cmd/mr_approve.go:66]: The `msg` var is not used when the user calls for a filename `-F`; shouldn't the quick action be added to the end of the file content when the user saves and closes it so he still get the "approve" action with the file content is used as the omment? (the same is valid for the unapprove code)

File: cmd/mr_unapprove.go
Comment (bmeneguele) [inline cmd/mr_unapprove.go:30]: Instead of creating this vars at the beginning, create them right before using them, like you did in the "approve" code, so can keep the code standard for both actions.

File: cmd/mr_unapprove.go
Comment (bmeneguele) [inline cmd/mr_unapprove.go:75]: The same situation I mentioned in the "approve" action.

File: cmd/note_common.go
Comment (bmeneguele) [inline cmd/note_common.go:288]: If a quick action is used with a comment the URL is still worthy showing, isn't it? It can be omitted only if no comments are given.

**Review by bmeneguele — CHANGES_REQUESTED** [review]

File: cmd/note_common.go
Comment (zampierilucas) [reply cmd/note_common.go:288]: The reason why I've added that is that in the case of _just_ using the quick action, the URL returned is `https://gitlab.com/lzampier/kernel-test/merge_requests/0#note_000000000`. 
It sure seems like a bug on GitLab's side, but since they treat quick actions as a different kind of `message`, I do understand they having nothing to point at.

Comment (bmeneguele) [reply cmd/note_common.go:291]: Hmm, got it. What if instead of then omitting the URL completely you check its suffix for the invalid `#note_000000000`?
I know it's not documented and might be a bug on their side and, possibly, get fixed one day. But if we get a test at least for this case we can make sure we it failing as soon as the behavior changes on their side.

I just really like the URLs left in the stdout after the user performs an action that effectively created a valid URL he can click on :). But feel free to disagree.

Comment (zampierilucas) [reply cmd/note_common.go:291]: Wait, first things first, the URL is __not__ shown only when the quick action is sent only thing sent, like:
```shell
$ lab mr approve 
Merge Request !1 approved
```
If there is a comment, the output is:
```shell
$ lab mr approve -m "Im a comment"
https://gitlab.com/lzampier/kernel-test/merge_requests/1#note_683541416
Merge Request !1 approved
```
Knowing that, do you still want me to show that  '000*' URL? btw I'm all in for that since I can pretty much drop that `quickaction bool` var, and make the createNote function a bit cleaner :D

Comment (bmeneguele) [reply cmd/note_common.go:291]: Ok, I miss understood part of the code! That's my fault.
The `quickaction` variable name in `createNote()` mislead me because it's only `true` when the user wants to also comment (`note == true`). So my thought was: "wait, but the approval without comments are also being done with quick actions, so the `quickactions` var will be `true`there as well."

Maybe changing the name of the variable or adding some comments there would be helpful to avoid this kind of back and forth to understand what's going on, not a hard requirement, just IMO :), your call. But yes, the code is correct!

Comment (zampierilucas) [reply cmd/note_common.go:291]: Yeah, thank you for pointing that out, and I agree the code is not particularly clear on that point, I will submit a follow-up patch.

**Review by bmeneguele — APPROVED** [review]
Thanks @zampierilucas :)

**Comment by zampierilucas** [comment]
I found an issue when using --with-comment

**Comment by zampierilucas** [comment]
Submitted a new commit with a fix to --with-comment and everything should be working fine.

**Comment by bmeneguele** [comment]
@zampierilucas I don't think the build issue you're facing is really related to your PR.
I just hit during the release testing: https://app.travis-ci.com/github/zaquestion/lab/builds/238658259

**Comment by zampierilucas** [comment]
> @zampierilucas I don't think the build issue you're facing is really related to your PR. I just hit during the release testing: https://app.travis-ci.com/github/zaquestion/lab/builds/238658259

@bmeneguele GREAT!! I mean bad, really bad :p
but you're right `mr_note_test.go:102` is the same one that is breaking on my MR, maybe it has been broken for a while?
OTOH I've tested the command that sends `lab mr reply origin 1:689060304 -m "text` and It succeedss without a problem, so are we having some issue with our testing <-> gitlab.com relation again?

**Comment by bmeneguele** [comment]
Possibly, but it shouldn't be panic'ing, but just dropping an error of any sort.
I'll try to take a look real quick to see if I can figure out and get the release out of the door.
However, if we can get it solved for the release, we can also get your PR also included before releasing :)

**Comment by bmeneguele** [comment]
The issue is related to something on gitlab side. Two commands in the very same MR, but in different `notes` behaves differently with no apparent reason:
```
$ lab mr reply lab-testing 3953:599577526 -m "test"
https://gitlab.com/lab-testing/test/merge_requests/3953#note_689563923
$ lab mr reply lab-testing 3953:689488483 -m "test"
$ 
```
The first run returns what is expected, in the last one, it doesn't.
For "old" notes, it seems to work just fine, while for new ones, it doesn't.

Something changed in the API?
I'm going to dig a further.

**Comment by bmeneguele** [comment]
For some really weird reason the discussions API is not returning new notes added to the MR, no matter if it's created using the Notes (where `individual_note = true`) or the Discussion API (where `individual_note = false`). 

I tested it using the API directly (through `curl`) so that's not an issue with `lab`, but definitely an issue on GitLab itself and a recent one, because the last PR that @prarit created was 4 days ago and everything was working just fine. I didn't find any issue open on gitlab.org though.

**Comment by bmeneguele** [comment]
Well, the problem turned out to be a lack of an HTTP header in API responses for MRs/Issues with more than 10k comments, which happens in our testing repo.

@zampierilucas helped me to find the issue and the PR https://github.com/zaquestion/lab/pull/745 was created with the fix.

**Comment by bmeneguele** [comment]
As soon as #745 is merged, this PR will need to be rebased.

**Comment by prarit** [comment]
10k?  What have you been doing to my MRs? :smile:

**Comment by zampierilucas** [comment]
@bmeneguele sure, I will rebase it as soon as 745 gets merged :D

> 10k? What have you been doing to my MRs? smile

@prarit GL considers 'entries' as any historical data on the MR, not only 'comments', for example, 'X person approved the MR' is also an entry, That's why we were able to beet GitLab 10k restriction that fast :p

**Comment by prarit** [comment]
Entries?  I'm just going to assume that somewhere there is a character-by-character review of one of my changesets.

**Comment by bmeneguele** [comment]
@zampierilucas #745 is now merged, please go ahead and rebase this one here so we can make sure it's also fine :)

**Review by bmeneguele — APPROVED** [review]

**Pull request merged**
